File: src/testapi.py
#!/usr/bin/env python3

import logging
import socket
import json

logger = logging.getLogger(__name__)


def upperKey(inParam):
    if type(inParam) is dict:
        return {k.upper():upperKey(v) for k,v in inParam.items()}
    elif type(inParam) is list:
        return [upperKey(i) for i in inParam]
    else:
        return inParam

def exactMatch(s):
    return f"^{s}$"

class CmdServer:
    def __init__(self):
        self.cmdUid = 1
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    def __del__(self):
        self.sock.close()

    def connect(self, address="localhost", port=8070):
        self.sock.connect((address, port))

    def sendCommand(self, jsonData):

        jsonData['UID'] = self.cmdUid
        self.cmdUid += 1

        # Send command
        msgStr = json.dumps(jsonData)
        msgBytes = msgStr.encode('utf-8')
        data = b'\xA5' + b'\xB1' + int.to_bytes(len(msgBytes), 4, 'big') + msgBytes 
        self.sock.sendall(data)

        # Read response
        rData = self.sock.recv(4096)
        return upperKey(json.loads(rData))


    def startTestServer(self, port=8090):
        jsonData = {'COMMAND': 'START_SERVER',
                    'COMMAND_DATA': { 'LPORT': port }  }
        resp = self.sendCommand(jsonData)
        return HttpServer(self, resp['COMMAND_DATA']['SERVER_ID'])


class HttpServer:
    def __init__(self, cmdSrv, id):
        self.cmdSrv = cmdSrv
        self.id = id

    def reset(self):
        jsonData = { 'COMMAND': 'RESET_SERVER',
                     'COMMAND_DATA': { 'SERVER_ID': self.id }  }
        resp = self.cmdSrv.sendCommand(jsonData)
    
    def kill(self):
        jsonData = { 'COMMAND': 'KILL_SERVER',
                     'COMMAND_DATA': { 'SERVER_ID': self.id }  }
        resp = self.cmdSrv.sendCommand(jsonData)

    def fetchStatus(self):
        jsonData = { 'COMMAND': 'FETCH_STATUS',
                     'COMMAND_DATA': { 'SERVER_ID': self.id }  }
        resp = self.cmdSrv.sendCommand(jsonData)
        return resp

    def checkStatus(self):
        allStatus = self.fetchStatus()['COMMAND_DATA']
        #print(json.dumps(allStatus, indent=2))

        msgLines = ['']
        for stat in allStatus:
            if 'UNMATCHED_RULES' in stat and stat['UNMATCHED_RULES']:
                msgLines.append( "Expected more request(s)...")
            if 'EXPECTED' in stat:
                if "REQUEST_INFO" in stat:
                    msgLines.append(f"Tried match the request '{stat['REQUEST_INFO']['METHOD']} {stat['REQUEST_INFO']['URI']}'...")
                for e in stat['EXPECTED']:
                    self.createExpectMessage(e, msgLines)
            msgLines.append("=====")
        
        return len(allStatus)==0, '\n'.join(msgLines)


    def createExpectMessage(self, sObj, msgLines, tab=0, tabSize=2):
        indent = " " * tabSize * tab
        if 'COLLECTION_TYPE' in sObj:
            if sObj['COLLECTION_TYPE'] == 'ALL_IN_ORDER':
                #msgLines.append(f"{indent}ALL_IN_ORDER")
                for c in sObj['COLLECTION']:
                    self.createExpectMessage(c, msgLines, tab+1, tabSize)
            elif sObj['COLLECTION_TYPE'] == 'ALL_IN_ANY_ORDER':
                #msgLines.append(f"{indent}ALL_IN_ANY_ORDER")
                self.createExpectMessage(sObj['COLLECTION'][0], msgLines, tab+1, tabSize)
                for c in sObj['COLLECTION'][1:]:
                    msgLines.append(f"{indent}== OR ==")
                    self.createExpectMessage(c, msgLines, tab+1, tabSize)
            else:
                raise Exception("Not implemented")
        else:        
            #print(sObj)
            for m in sObj['RULE']:
                msgLines.append(f"{indent}{m}")


    def expect(self, ruleOrCollection):
        jsonData = { 'COMMAND': 'ADD_RULE',
                    'COMMAND_DATA': { 'SERVER_ID': self.id, 'RULE': ruleOrCollection.toJson() }  }
        resp = self.cmdSrv.sendCommand(jsonData)


        
class Response:
    def __init__(self, respData="", statusCode=200):
        self.statusCode = statusCode
        self.respHeaders = {}
        self.respData = respData

    def toJson(self):
        retObj = {'CODE': self.statusCode}
        if self.respHeaders:
            retObj['HEADERS'] = self.respHeaders
        if self.respData:
            retObj['DATA'] = self.respData
        return retObj

    def code(self, code):
        self.statusCode = code
        return self

    def headers(self, headers):
        self.respHeaders = headers
        return self

    def data(self, data):
        self.respData = data
        return self


class Matcher:
    URL = "URL"
    METHOD = "METHOD"
    HEADER = "HEADER"
    DATA = "DATA"
    FILE_DATA = "FILE_DATA"

    def __init__(self, entity, matchValue):
        self.entity = entity
        self.matchValue = matchValue
        self.negate = False

    def toJson(self):
        return { 'TYPE': self.entity,
                 'VALUE': self.matchValue,
                 'NEGATE': self.negate
        }
    

class Rule:
    def __init__(self):
        self.matchers = []
        self.calledAtLeast = 1
        self.calledAtMost = 100000
        self.response = None

    def toJson(self):
        return {
            'TYPE': 'MATCHER',
            'MATCHERS': [m.toJson() for m in self.matchers],
            'CALLED_TIMES': {'AT_LEAST': self.calledAtLeast, 'AT_MOST': self.calledAtMost},
            'RESPONSE': self.response.toJson()
        }

    def matchTimes(self, t):
        self.calledAtLeast = t
        self.calledAtMost = t
        return self

    def matchAtLeast(self, t):
        self.calledAtLeast = t
        return self

    def matchAtMost(self, t):
        self.calledAtMost = t
        return self

    def matchTimesRange(self, t1, t2):
        self.calledAtLeast = t1
        self.calledAtMost = t2
        return self

    def method(self, method):
        m = Matcher(Matcher.METHOD, method)
        self.matchers.append(m)
        return self


    def url(self, url):
        m = Matcher(Matcher.URL, url)
        self.matchers.append(m)
        return self

    def header(self, headerKey, headerVal):
        m = Matcher(Matcher.HEADER, (headerKey, headerVal))
        self.matchers.append(m)
        return self

    def headers(self, headers):
        for k in headers:
            m = Matcher(Matcher.HEADER, (k,headers[k]))
            self.matchers.append(m)
        return self

    def data(self, data):
        m = Matcher(Matcher.DATA, data)
        self.matchers.append(m)
        return self

    def fileData(self, filePath):
        m = Matcher(Matcher.FILE_DATA, filePath)
        self.matchers.append(m)
        return self

    def respondWith(self, r):
        self.response = r
        return self


class Collection:
    ALL_IN_ORDER = "ALL_IN_ORDER"
    ALL_IN_ANY_ORDER = "ALL_IN_ANY_ORDER"
    ANY_NUMBER = "ANY_NUMBER"

    def __init__(self):
        self.type = Collection.ALL_IN_ORDER
        self.times = 1 
        self.anyNum = 1
        self.rules = []

    def toJson(self):
        jsonObj = {
            'TYPE': 'COLLECTION',
            'COLLECTION_TYPE': self.type,
            'CALLED_TIMES': self.times,
            'RULES': [ r.toJson() for r in self.rules]
            }

        if self.type == Collection.ANY_NUMBER:
            jsonObj['ANY_NUMBER'] = self.anyNum

        return jsonObj


    def expectAllInOrder(self):
        self.type = Collection.ALL_IN_ORDER
        return self

    def expectAllInAnyOrder(self):
        self.type = Collection.ALL_IN_ANY_ORDER
        return self

    def expectAnyNumber(self, num):
        self.type = Collection.ANY_NUMBER
        self.anyNum = num
        return self
    
    def addRule(self, rule):
        self.rules.append(rule)
        return self

   

if __name__ == "__main__":
    pass

File: bindings/c_sharp/testapi/testapi.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Collections.Generic;

using Newtonsoft.Json;
using Newtonsoft.Json.Linq;


namespace testapi
{
    public class CommandClient
    {
        private uint m_uid = 1;
        private string m_host = "";
        private int m_hostPort = 0;
        private TcpClient m_client;
        private NetworkStream m_sockStream;

        public bool Connect(string host="localhost", int port=8070)
        {          
            Console.WriteLine("Connect");

            try
            {            
                m_client = new TcpClient(host, port);
                m_sockStream = m_client.GetStream();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
                return false;
            }

            m_host = host;
            m_hostPort = port;


            return true;
        }

        public void Disconnect(bool killHttpServers=true)
        {
            if (killHttpServers)
            {
                JObject jObj = new JObject();
                SendCommand("KILL_ALL_SERVERS", jObj);                
            }
            

            Console.WriteLine("Disconnect");
            m_sockStream.Close();
            m_client.Close();
        }

        public HttpServer StartTestServer(int port=8090)
        {
            var jObj = new JObject();
            jObj.Add(new JProperty("LPORT", port));
            var resp = SendCommand("START_SERVER", jObj);           

            return new HttpServer(this, (int)resp.SelectToken("COMMAND_DATA.SERVER_ID"));
        }

        public JObject SendCommand(string command, JObject commandData)
        {
            JObject jObj = new JObject();
            jObj.Add(new JProperty("UID", m_uid));
            m_uid++;
            jObj.Add(new JProperty("COMMAND", command));
            jObj.Add(new JProperty("COMMAND_DATA", commandData));

            string jsonStr = jObj.ToString();


            byte[] jsonData = Encoding.UTF8.GetBytes(jsonStr);
            byte[] headerData = { 0xA5, 0xB1 };
            byte[] lengthData = BitConverter.GetBytes(jsonData.Length);
            if (BitConverter.IsLittleEndian)
                Array.Reverse(lengthData);

            m_sockStream.Write(headerData, 0, headerData.Length);
            m_sockStream.Write(lengthData, 0, lengthData.Length);
            m_sockStream.Write(jsonData, 0, jsonData.Length);

            byte[] bytes = new byte[4096];
            int bytesRead = m_sockStream.Read(bytes, 0, bytes.Length);

            jsonStr = Encoding.UTF8.GetString(bytes);

            JObject resp = JObject.Parse(jsonStr);

            //Console.WriteLine(resp["COMMAND_DATA"]["SERVER_ID"]);
            //Console.WriteLine(resp.SelectToken("COMMAND_DATA.SERVER_ID"));
            //Console.WriteLine(resp);

            return resp;

        }

    }



    public class HttpServer
    {
        protected CommandClient m_owner;
        protected int m_id;

        public HttpServer(CommandClient owner, int id)
        {
            m_owner = owner;
            m_id = id;
        }

        ~HttpServer()
        {
            Kill();
        }


        public void Reset()
        {
            JObject jObj = new JObject();
            jObj.Add(new JProperty("SERVER_ID", m_id));
            m_owner.SendCommand("RESET_SERVER", jObj);
        }

        public void Kill()
        {
            JObject jObj = new JObject();
            jObj.Add(new JProperty("SERVER_ID", m_id));
            m_owner.SendCommand("KILL_SERVER", jObj);
        }

        public JObject FetchStatus()
        {
            JObject jObj = new JObject();
            jObj.Add(new JProperty("SERVER_ID", m_id));
            return m_owner.SendCommand("FETCH_STATUS", jObj);
        }

        public bool CheckStatus(out string msg)
        {
            var jObj = FetchStatus();
            var allStatus = jObj["COMMAND_DATA"];

            //Console.WriteLine(allStatus.ToString());

            List<string> msgLines = new List<string>();
            foreach (var stat in allStatus)
            {

                var elem = stat.ToObject<JObject>();
                if (elem.ContainsKey("UNMATCHED_RULES") && (bool)elem.SelectToken("UNMATCHED_RULES"))
                {
                    msgLines.Add("Expected more request(s)...");
                }

                if (elem.ContainsKey("EXPECTED"))
                {
                    if (elem.ContainsKey("REQUEST_INFO"))
                    {
                        var reqInfo = elem["REQUEST_INFO"].ToObject<JObject>();
                        msgLines.Add($"Tried match the request '{reqInfo["METHOD"]} {reqInfo["URI"]}'...");
                    }

                    foreach(var e in stat["EXPECTED"])
                    {
                        var expectObj = e.ToObject<JObject>();
                        CreateExpectMessage(expectObj, msgLines);
                    }
                }
                msgLines.Add("=====");
            }


            msg = String.Join("\n", msgLines);

            return false;
        }

        protected void CreateExpectMessage(JObject obj, List<string> msgLines, int tab=0, int tabSize=2)
        {
            string indent = new string(' ', tab * tabSize);
            
            if (obj.ContainsKey("COLLECTION_TYPE"))
            {
                // Collection...
                string colType = (string)obj["COLLECTION_TYPE"];
                if (colType == "ALL_IN_ORDER")
                {
                    foreach (var c in obj["COLLECTION"])
                    {
                        CreateExpectMessage(c.ToObject<JObject>(), msgLines, tab+1, tabSize);
                    }

                }
                else if (colType == "ALL_IN_ANY_ORDER")
                {
                    msgLines.Add($"{indent}== ALL_IN_ANY_ORDER collection ==");
                    var arr = (JArray)obj["COLLECTION"];
                    CreateExpectMessage(arr[0].ToObject<JObject>(), msgLines, tab+1, tabSize);
                    for(var i=1; i<arr.Count; ++i)
                    {
                        msgLines.Add($"{indent}== OR ==");
                        CreateExpectMessage(arr[i].ToObject<JObject>(), msgLines, tab+1, tabSize);
                    }
                }
                else if (colType == "ANY_NUM")
                {
                    msgLines.Add($"{indent}== ANY_NUM collection ==");
                    var arr = (JArray)obj["COLLECTION"];
                    CreateExpectMessage(arr[0].ToObject<JObject>(), msgLines, tab+1, tabSize);
                    for(var i=1; i<arr.Count; ++i)
                    {
                        msgLines.Add($"{indent}== OR ==");
                        CreateExpectMessage(arr[i].ToObject<JObject>(), msgLines, tab+1, tabSize);
                    }
                }
            }
            else
            {
                // Rule(s)
                foreach(var m in obj["RULE"])
                {
                    msgLines.Add($"{indent}{(string)m}");
                }
            }
        }

        public void Expect(ExpectBase eb)
        {
            var jObj = new JObject();
            jObj.Add(new JProperty("SERVER_ID", m_id));
            jObj.Add(new JProperty("RULE", eb.ToJson()));            
            m_owner.SendCommand("ADD_RULE", jObj);
        }


    }


    public class Response
    {
        protected int m_statusCode;
        protected Dictionary<string, string>  m_respHeaders; 
        protected string m_respData;


        public Response(string respData="", int statusCode=200)
        {
            m_statusCode = statusCode;
            m_respHeaders = new Dictionary<string, string>();
            m_respData = respData;
        }

        public JObject ToJson()
        {
            var jObj = new JObject();
            jObj.Add(new JProperty("CODE", m_statusCode));
            jObj.Add(new JProperty("HEADERS", JObject.FromObject(m_respHeaders)));
            jObj.Add(new JProperty("DATA", m_respData));
            return jObj;        
        }        

        public Response Code(int statusCode)
        {
            m_statusCode = statusCode;
            return this;            
        }

        public Response Headers(Dictionary<string, string> headers)
        {
            m_respHeaders = headers;
            return this;            
        }

        public Response Data(string respData)
        {
            m_respData = respData;
            return this;            
        }
    }


    public class Matcher
    {
        public enum MatchType { URL, METHOD, HEADER, DATA, FILE_DATA };
        
        protected MatchType m_matchType;
        protected string m_matchValue1;
        protected string m_matchValue2; // So far, only used with header matching since it is key/value matching
        protected bool m_negate = false;

        public Matcher(MatchType mt, string matchValue, string matchValue2="")
        {
            m_matchType = mt;
            m_matchValue1 = matchValue;
            m_matchValue2 = matchValue2;
        }

        public JObject ToJson()
        {
            var jObj = new JObject();
            jObj.Add(new JProperty("TYPE", m_matchType.ToString()));

            if (m_matchType != MatchType.HEADER)
            {
                jObj.Add(new JProperty("VALUE", m_matchValue1));
            }
            else
            {
                var hArr = new JArray();
                hArr.Add(m_matchValue1);
                hArr.Add(m_matchValue2);
                jObj.Add(new JProperty("VALUE", hArr));
            }

            jObj.Add(new JProperty("NEGATE", m_negate));
            return jObj;        
        }        
    }

    public abstract class ExpectBase
    {
        protected enum BaseType { NONE, RULE, COLLECTION };
        protected BaseType baseType;

        public abstract JObject ToJson();
    }

    public class Rule : ExpectBase
    {
        protected List<Matcher> m_matchers = new List<Matcher>(); 
        protected int m_calledAtLeast = 1;
        protected int m_calledAtMost = 100000;
        protected Response m_response;        

        public Rule()
        {
            baseType = BaseType.RULE;
        }

        public override JObject ToJson()
        {
            var jObj = new JObject();

            jObj.Add(new JProperty("TYPE", "MATCHER"));

            var jArrMatch = new JArray();
            foreach(var m in m_matchers)
            {
                jArrMatch.Add(m.ToJson());
            }

            jObj.Add(new JProperty("MATCHERS", jArrMatch));

            var jObjCalled = new JObject();
            jObjCalled.Add(new JProperty("AT_LEAST", m_calledAtLeast));
            jObjCalled.Add(new JProperty("AT_MOST", m_calledAtMost));
            jObj.Add(new JProperty("CALLED_TIMES", jObjCalled));

            jObj.Add(new JProperty("RESPONSE", m_response.ToJson()));

            return jObj;        
        }


        public Rule MatchTimes(int times)
        {
            m_calledAtLeast = times;
            m_calledAtMost = times;
            return this;
        }


        public Rule MatchAtLeast(int times)
        {
            m_calledAtLeast = times;
            return this;
        }

        public Rule MatchAtMost(int times)
        {
            m_calledAtMost = times;
            return this;
        }

        public Rule MatchTimesRange(int t1, int t2)
        {
            m_calledAtLeast = t1;
            m_calledAtMost = t2;
            return this;
        }

        public Rule Method(string method)
        {
            var m = new Matcher(Matcher.MatchType.METHOD, method);
            m_matchers.Add(m);
            return this;
        }

        public Rule Url(string url)
        {
            var m = new Matcher(Matcher.MatchType.URL, url);
            m_matchers.Add(m);
            return this;
        }


        public Rule Header(string headerName, string headerValue)
        {
            var m = new Matcher(Matcher.MatchType.HEADER, headerName, headerValue);
            m_matchers.Add(m);
            return this;
        }

        public Rule Date(string data)
        {
            var m = new Matcher(Matcher.MatchType.DATA, data);
            m_matchers.Add(m);
            return this;
        }

        public Rule RespondWith(Response response)
        {
            m_response = response;
            return this;
        }

    }

    public class Collection : ExpectBase
    {
        protected enum CollectionType { ALL_IN_ORDER, ALL_IN_ANY_ORDER, ANY_NUMBER };
        protected CollectionType m_type = CollectionType.ALL_IN_ORDER;
        protected int m_times = 1;
        protected int m_maxNum = 1;
        protected List<Rule> m_rules = new List<Rule>();

        public Collection()
        {
            baseType = BaseType.COLLECTION;
        }

        public override JObject ToJson()
        {
            var jObj = new JObject();
            jObj.Add(new JProperty("TYPE", "COLLECTION"));
            jObj.Add(new JProperty("COLLECTION_TYPE", m_type.ToString()));
            jObj.Add(new JProperty("CALLED_TIMES", m_times));
            jObj.Add(new JProperty("MAX_NUMBER", m_times));

            var jArrRules = new JArray();
            foreach(var r in m_rules)
            {
                jArrRules.Add(r.ToJson());
            }
            jObj.Add(new JProperty("RULES", jArrRules));

            return jObj;        
        }

        public Collection ExpectAllInOrder()
        {
            m_type = CollectionType.ALL_IN_ORDER;
            return this;
        }

        public Collection ExpectAllInAnyOrder()
        {
            m_type = CollectionType.ALL_IN_ANY_ORDER;
            return this;
        }

        public Collection ExpectAnyNumber(int maxNum)
        {
            m_type = CollectionType.ANY_NUMBER;
            m_maxNum = maxNum;
            return this;
        }

        public Collection AddRule(Rule r)
        {
            m_rules.Add(r);
            return this;
        }
    }
}